### PR TITLE
Automation coverage for BZ 2053843

### PR DESCRIPTION
**Pull request by Gauravtalreja1**
**Description:**
1. Adding automation coverage for [BZ 2053843](https://bugzilla.redhat.com/show_bug.cgi?id=2053843),
2. Some fixes to use target_sat's cli/cli_factory
3. Using request.addfinalizer instead of try-finally blocks

**Test Results:**
```
(.robottelo) ➜  robottelo git:(hammer-shell) ✗ pytest -q --disable-pytest-warnings tests/foreman/cli/test_hammer.py -k test_positive_hammer_shell
.                                                                                                                                                     [100%]
1 passed, 3 deselected, 2 warnings in 38.74s
```

### CODE REVIEW

File: tests/foreman/cli/test_hammer.py [hunk 1]
```diff
@@ -12,28 +12,25 @@
...
-from robottelo.cli.factory import make_product
 from robottelo.constants import DataFile
+from robottelo.logging import logger
 
 
-HAMMER_COMMANDS = json.loads(DataFile.HAMMER_COMMANDS_JSON.read_bytes())
+HAMMER_COMMANDS = json.loads(DataFile.HAMMER_COMMANDS_JSON.read_text())
 
-pytestmark = [pytest.mark.tier1]
+pytestmark = [pytest.mark.tier1, pytest.mark.upgrade]
```
Comment (jyejare) [inline tests/foreman/cli/test_hammer.py:33]: add e2e as well ?

File: tests/foreman/cli/test_hammer.py [hunk 2]
```diff
@@ -144,33 +136,32 @@ def test_positive_disable_hammer_defaults(target_sat):
...
-        assert default_product_name in result.stdout
-    finally:
-        Defaults.delete({'param-name': 'organization_id'})
+    default_org = target_sat.cli_factory.make_org()
+    default_product = target_sat.cli_factory.make_product({'organization-id': default_org.id})
+
+    target_sat.cli.Defaults.add({'param-name': 'organization_id', 'param-value': default_org.id})
+    # list templates for BZ#1368173
+    result = target_sat.execute('hammer job-template list')
+    assert result.status == 0
```
Comment (jyejare) [inline tests/foreman/cli/test_hammer.py:147]: This does not simply confirms if there are any templates listed or empty list returned. So I would check the output and the list is not empty.

File: tests/foreman/cli/test_hammer.py [hunk 3]
```diff
@@ -144,33 +136,32 @@ def test_positive_disable_hammer_defaults(target_sat):
...
+    # Verify --organization-id is not required to pass if defaults are set
+    result = target_sat.execute('hammer product list')
+    assert result.status == 0
```
Comment (jyejare) [inline tests/foreman/cli/test_hammer.py:150]: Same here!

**Review by jyejare — CHANGES_REQUESTED** [review]
ACK with minor comments!

File: tests/foreman/cli/test_hammer.py (hunk 2 above)
Comment (Gauravtalreja1) [reply tests/foreman/cli/test_hammer.py:147]: Here we just would like to confirm that after specifying defaults param, hammer doesn't asks to pass org/orgid

File: tests/foreman/cli/test_hammer.py (hunk 3 above)
Comment (Gauravtalreja1) [reply tests/foreman/cli/test_hammer.py:150]: Same as above

File: tests/foreman/cli/test_hammer.py (hunk 1 above)
Comment (Gauravtalreja1) [reply tests/foreman/cli/test_hammer.py:33]: There are not e2e tests, still can add?

File: tests/foreman/cli/test_hammer.py [hunk 4]
```diff
@@ -144,33 +136,32 @@ def test_positive_disable_hammer_defaults(target_sat):
...
-        assert result.status != 0
-        assert default_product_name not in result.stdout
-        # Verify --organization-id is not required to pass if defaults are set
-        result = target_sat.execute('hammer --use-defaults product list')
-        assert result.status == 0
-        assert default_product_name in result.stdout
-    finally:
-        Defaults.delete({'param-name': 'organization_id'})
+    default_org = target_sat.cli_factory.make_org()
+    default_product = target_sat.cli_factory.make_product({'organization-id': default_org.id})
```
Comment (Griffin-Sullivan) [inline tests/foreman/cli/test_hammer.py:142]: Could you use `module_org` and `module_product` here?

**Review by Griffin-Sullivan — APPROVED** [review]

File: tests/foreman/cli/test_hammer.py (hunk 4 above)
Comment (Gauravtalreja1) [reply tests/foreman/cli/test_hammer.py]: Sure, we can, updated the test to use `function_product` fixture, thank you @Griffin-Sullivan

File: tests/foreman/cli/test_hammer.py (hunk 1 above)
Comment (Gauravtalreja1) [reply tests/foreman/cli/test_hammer.py]: @jyejare Updated

Comment (pondrejk) [reply tests/foreman/cli/test_hammer.py]: imho `test_positive_all_options` falls into upgrade/e2e category, but for other tests in the module I'm not convinced

Comment (Gauravtalreja1) [reply tests/foreman/cli/test_hammer.py]: @pondrejk As this are all basic hammer scenarios, so i think its fine to include those into upgrade alltier and e2e runs?

**Review by jyejare — APPROVED** [review]

**Pull request merged**